src/shared_queue.cc

```diff
@@ -21,23 +21,73 @@
 #include <unordered_map>
 #include <unordered_set>
 
+#include "include/proxy-wasm/vm_id_handle.h"
+
 namespace proxy_wasm {
 
 SharedQueue &getGlobalSharedQueue() {
   static auto *ptr = new SharedQueue;
   return *ptr;
 }
 
+SharedQueue::SharedQueue(bool register_vm_id_callback) {
+  if (register_vm_id_callback) {
+    registerVmIdHandleCallback([this](std::string_view vm_id) { this->deleteByVmId(vm_id); });
+  }
+}
+
+void SharedQueue::deleteByVmId(std::string_view vm_id) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto queue_keys = vm_queue_keys_.find(std::string(vm_id));
+  if (queue_keys != vm_queue_keys_.end()) {
+    for (std::size_t queue_key : queue_keys->second) {
+      auto token = queue_tokens_.find(queue_key);
+      if (token != queue_tokens_.end()) {
+        queues_.erase(token->second);
+        queue_tokens_.erase(token);
+      }
+    }
+    vm_queue_keys_.erase(queue_keys);
+  }
+}
+
+uint32_t SharedQueue::nextQueueToken() {
+  // TODO(@mathetake): Should we handle the case where the queue overflows, i.e. the number of used
+  // tokens exceeds the max of uint32? If it overflows, the following loop never exits.
+  while (true) {
+    uint32_t token = next_queue_token_++;
+    if (token == 0) {
+      continue; // 0 is an illegal token.
+    }
+
+    if (queues_.find(token) == queues_.end()) {
+      return token;
+    }
+  }
+}
+
 uint32_t SharedQueue::registerQueue(std::string_view vm_id, std::string_view queue_name,
                                     uint32_t context_id, CallOnThreadFunction call_on_thread,
                                     std::string_view vm_key) {
   std::lock_guard<std::mutex> lock(mutex_);
-  auto key = std::make_pair(std::string(vm_id), std::string(queue_name));
+
+  // TODO(mathetake): care about hash collision?
+  auto key = queueKeyHash(vm_id, queue_name);
```

Comment:
Nit: What's the reason for the change from automatic to manual calculation of the hash (using `queueKeyHash` instead of allowing `std::unordered_map` to calculate hash of the key)?

Reply:
I thought it was necessary somehow, but I realize now it's not. Do you want me to roll back? I'm ok with that.

Reply:
I think it makes things a bit more confusing (what's this special hash? where is `size_t` coming from?), so it's not an improvement in my mind, but if you prefer to leave it, then we can leave it. You're the one who's going to maintain this anyway :)

Reply:
> a bit more confusing 

That's true. I don't have preference so I will revert the change :+1: